File: STSecurityTests/STSecurityKeychainAccessPasswordTests.m
//
//  STSecurityKeychainAccessPasswordTests.m
//  STSecurity
//
//  Copyright (c) 2012 Scott Talbot. All rights reserved.
//

@import XCTest;

#import "STSecurityKeychainAccess.h"


static NSString * const service = @"STSecurityTest";


@interface STSecurityKeychainAccessPasswordTests : XCTestCase
@end

@implementation STSecurityKeychainAccessPasswordTests

- (void)setUp {
	[STSecurityKeychainAccess deletePasswordsForService:service];
}

- (void)testFetchNonexistent {
	NSString * const username = @"username";

	{
		NSError *error = nil;
		NSString * const password = [STSecurityKeychainAccess passwordForUsername:username service:service error:&error];
		XCTAssertNil(password);
		XCTAssertNotNil(error);
	}
}

- (void)testRoundtrip {
	NSString * const username = @"username";
	NSString * const password = @"password";

	{
		NSError *error = nil;
		BOOL const status = [STSecurityKeychainAccess setPassword:password forUsername:username service:service error:&error];
		XCTAssertTrue(status);
		XCTAssertNil(error);
	}

	{
		NSError *error = nil;
		NSString * const fetchedPassword = [STSecurityKeychainAccess passwordForUsername:username service:service error:&error];
		XCTAssertNotNil(fetchedPassword);
		XCTAssertNil(error, @"error: %@", error);
		XCTAssertEqualObjects(password, fetchedPassword);
	}
}

- (void)testMultipleInsertion {
	NSString * const username = @"username";
	NSString * const password = @"password";

	{
		NSError *error = nil;
		BOOL const status = [STSecurityKeychainAccess setPassword:password forUsername:username service:service error:&error];
		XCTAssertTrue(status);
		XCTAssertNil(error);
	}

	{
		NSError *error = nil;
		BOOL const status = [STSecurityKeychainAccess setPassword:password forUsername:username service:service error:&error];
		XCTAssertFalse(status);
		XCTAssertNotNil(error, @"error: %@", error);
	}

	{
		STSecurityKeychainWritingOptions * const options = [[STSecurityKeychainWritingOptions alloc] init];
		options.accessibility = STSecurityKeychainItemAccessibleAlways;
		options.overwriteExisting = YES;

		NSError *error = nil;
		BOOL const status = [STSecurityKeychainAccess setPassword:password forUsername:username service:service withOptions:options error:&error];
		XCTAssertTrue(status);
		XCTAssertNil(error, @"error: %@", error);
	}

	{
		NSError *error = nil;
		BOOL const status = [STSecurityKeychainAccess setPassword:password forUsername:username service:service error:&error];
		XCTAssertFalse(status);
		XCTAssertNotNil(error, @"error: %@", error);
	}

	{
		STSecurityKeychainWritingOptions * const options = [[STSecurityKeychainWritingOptions alloc] init];
		options.accessibility = STSecurityKeychainItemAccessibleAlwaysThisDeviceOnly;
		options.overwriteExisting = YES;

		NSError *error = nil;
		BOOL const status = [STSecurityKeychainAccess setPassword:password forUsername:username service:service withOptions:options error:&error];
		XCTAssertTrue(status);
		XCTAssertNil(error, @"error: %@", error);
	}

	{
		NSError *error = nil;
		NSString * const fetchedPassword = [STSecurityKeychainAccess passwordForUsername:username service:service error:&error];
		XCTAssertNotNil(fetchedPassword);
		XCTAssertNil(error, @"error: %@", error);
		XCTAssertEqualObjects(password, fetchedPassword);
	}

	{
		NSError *error = nil;
		BOOL const status = [STSecurityKeychainAccess deletePasswordForUsername:username service:service error:&error];
		XCTAssertTrue(status);
		XCTAssertNil(error, @"error: %@", error);
	}
}

@end

File: STSecurityTests/STSecurityKeychainAccessPasswordTests.swift
//  Copyright (c) 2014 Scott Talbot. All rights reserved.

import XCTest


let service = "STSecurityTest"


class STSecurityKeychainAccessPasswordSwiftTests: XCTestCase {

	override func setUp() {
		do {
			try STSecurityKeychainAccess.deletePasswords(service: service)
		} catch {}
	}

	func testFetchNonexistent() {
		let username = "username"

		var password: String?
		do {
			password = try STSecurityKeychainAccess.password(service: service, username: username)
			XCTAssertTrue(false)
		} catch (let error) {
			XCTAssertNotNil(error)
		}
		XCTAssertNil(password)
	}

	func testRoundtrip() {
		let username = "username"
		let password = "password"

		do {
			try STSecurityKeychainAccess.setPassword(service: service, username: username, password: password)
		} catch (let error) {
			XCTAssertNil(error)
		}

		var fetchedPassword: String?
		do {
			fetchedPassword = try STSecurityKeychainAccess.password(service: service, username: username)
		} catch (let error) {
			XCTAssertTrue(false)
			XCTAssertNotNil(error)
		}
		XCTAssertNotNil(fetchedPassword)
		XCTAssertEqual(password, fetchedPassword)
	}

	func testMultipleInsertion() {
		let username = "username"
		let password = "password";

		do {
			try STSecurityKeychainAccess.setPassword(service: service, username: username, password: password)
		} catch (let error) {
			XCTAssertTrue(false)
			XCTAssertNotNil(error)
		}

		do {
			try STSecurityKeychainAccess.setPassword(service: service, username: username, password: password)
			XCTAssertTrue(false)
		} catch (let error) {
			XCTAssertNotNil(error)
		}

		do {
			try STSecurityKeychainAccess.setPassword(service: service, username: username, password: password, options: STSecurityKeychainWritingOptions(overwriteExisting: true,accessibility: .accessibleAlways))
		} catch (let error) {
			XCTAssertTrue(false)
			XCTAssertNotNil(error)
		}

		do {
			try STSecurityKeychainAccess.setPassword(service: service, username: username, password: password)
			XCTAssertTrue(false)
		} catch (let error) {
			XCTAssertNotNil(error)
		}

		do {
			try STSecurityKeychainAccess.setPassword(service: service, username: username, password: password, options: STSecurityKeychainWritingOptions(overwriteExisting: true, accessibility: .accessibleAlwaysThisDeviceOnly))
		} catch (let error) {
			XCTAssertTrue(false)
			XCTAssertNotNil(error)
		}

		var fetchedPassword: String?
		do {
			fetchedPassword = try STSecurityKeychainAccess.password(service: service, username: username)
		} catch (let error) {
			XCTAssertTrue(false)
			XCTAssertNotNil(error)
		}
		XCTAssertNotNil(fetchedPassword);
		XCTAssertEqual(password, fetchedPassword);

		do {
			try STSecurityKeychainAccess.deletePassword(service: service, username: username)
		} catch (let error) {
			XCTAssertTrue(false)
			XCTAssertNotNil(error)
		}
	}

}
